[Rocprofiler-systems] : Add XGMI and PCIe metrics to the profiling data

sputhala-amd opened the pull request:
## Motivation

Add XGMI and PCIE data for GPU Interconnect Traffic Metrics Collection

## Technical Details

Add support for AMD XGMI (GPU-to-GPU interconnect) and PCIe metrics:
  * XGMI link width in bits
  * XGMI link speed in GT/s
  * Per-link read bandwidth (KB)
  * Per-link write bandwidth (KB)

-  PCIe metrics:
  * PCIe link width
  * PCIe link speed in GT/s
  * Accumulated bandwidth (MB)
  * Instantaneous bandwidth (MB/s)

## Test Plan

Manual tests performed. Test Automation will be added soon for validation in ctest.

## Test Result

XGMI and PCIE tracks are confirmed in perfetto and Visualizer.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

sputhala-amd committed:
Add XGMI and PCIe metrics to the profiling data

Add support for AMD XGMI (GPU-to-GPU interconnect) and PCIe
metrics:
  * XGMI link width in bits
  * XGMI link speed in GT/s
  * Per-link read bandwidth (KB)
  * Per-link write bandwidth (KB)

- Add new categories for PCIe metrics:
  * PCIe link width
  * PCIe link speed in GT/s
  * Accumulated bandwidth (MB)
  * Instantaneous bandwidth (MB/s)

sputhala-amd commented:
Running with TransferBench:

```
Summary: [268435456 bytes per Transfer] [GFX:8] [1 Read(s) 1 Write(s)]
===========================================================================
SRC\DST  GPU 00     GPU 01     GPU 02     GPU 03     GPU 04     GPU 05     GPU 06     GPU 07        STotal      Actual
GPU 00      N/A     47.764     49.169     48.625     46.748     47.991     47.579     47.407       335.284     327.236
GPU 01   47.667        N/A     47.890     48.860     48.397     46.739     47.636     47.222       334.412     327.175
GPU 02   48.833     47.920        N/A     47.991     48.928     48.660     46.749     48.105       337.186     327.243
GPU 03   48.505     49.060     48.000        N/A     48.032     48.811     48.455     46.977       337.841     328.842
GPU 04   46.986     48.659     48.628     48.040        N/A     48.121     48.922     48.850       338.207     328.903
GPU 05   47.919     46.686     48.521     48.845     48.059        N/A     47.746     48.566       336.342     326.802
GPU 06   47.649     47.480     47.153     48.298     49.259     47.986        N/A     47.862       335.688     330.073
GPU 07   47.315     47.443     47.823     46.509     48.831     48.794     47.965        N/A       334.679     325.560
```

<img width="1600" height="584" alt="image" src="https://github.com/user-attachments/assets/65dbf36f-9cb9-4f06-989b-07e3839cdbdc" />

<img width="2134" height="1342" alt="image" src="https://github.com/user-attachments/assets/483465aa-5bd0-4cb8-bd64-e98a8cf7f01f" />

sputhala-amd commented:
@mradosav-amd Can you review the changes to the caching in amd-smi samples.

sputhala-amd committed:
Fix VCN/JPEG insert logic

mradosav-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_post_processing.cpp:
```diff
@@ -536,56 +574,143 @@ rocpd_post_processing::get_amd_smi_sample_callback() const
-        for(size_t idx = 0; idx < xcp_metrics.size(); ++idx)
+        // Process uint16_t vector-based metrics
+        // Check if VCN activity is supported and insert accordingly
+        if(gpu::is_vcn_activity_supported(_amd_smi.device_id))
```
I guess you don't need gpu included at all because you are already passing that here:
```
        bool is_vcn_enabled  = settings_bits.test(static_cast<int>(pos::vcn_activity));
        bool is_jpeg_enabled = settings_bits.test(static_cast<int>(pos::jpeg_activity));
        bool is_xgmi_enabled = settings_bits.test(static_cast<int>(pos::xgmi));
        bool is_pcie_enabled = settings_bits.test(static_cast<int>(pos::pcie));
```

sputhala-amd replied:
`is_vcn_enabled` enabled captures the settings passed by the user. We need `gpu::is_vcn_activity_supported` in order to check whether we need to insert the `vcn_activity` values or the `xcp.vcn_busy` values.
The data handling and track naming for xcp and activity values are different. Hence we need the check  `gpu::is_vcn_activity_supported` to distinguish the 2 different data format.

mradosav-amd replied:
But, `rocpd_post_processing` doesn't need to know about gpu info, it's just inserting data into rocpd. We can do the preparation of data in `amd_smi` and just send it to `rocpd_post_processing` via `trace_cache`. Inside `amd_smi` we can decide should we use `vcn_activity` or `xcp.vcn_busy`, it's basically the same data. `rocpd_post_processing` can just decide should it write it or not, via settings value.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_post_processing.cpp:
```diff
@@ -405,71 +405,107 @@ rocpd_post_processing::get_pmc_event_with_sample_callback() const
 rocpd_post_processing::get_amd_smi_sample_callback() const
 {
-    struct xcp_metrics_t
+#if ROCPROFSYS_USE_ROCM > 0
```
Why do we need this ifdef?

sputhala-amd replied:
We actually need the ifdef only for  `gpu::is_vcn_activity_supported` since this flag is defined inside gpu.cpp only with this ifdef after checking ROCM availability.

However I thought we cannot process any amd_smi data if ROCM is not available. I looked at the function `rocpd_post_processing::register_parser_callback` where we register this callback and the whole block is within the same ifdef for ROCM.

mradosav-amd replied:
Sure, idea was the `rocpd_post_processing` code is minimally dependent on other modules, it is module for its own. I would not introduce new dependency there if it is not necessary.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi.cpp:
```diff
@@ -335,67 +447,91 @@ get_state()
 }
 
 std::vector<uint8_t>
```
Isn't it more simpler to serialize all the data unconditinally? We are passing what is enabled/disabled or supported/not supported, and we will just write that inside rocpd. In that way, `rocpd_post_processing` will not be depentet on gpu.hpp. Idea is to have basically no dependencies inside `rocpd_post_processing`, everything should be passed via trace_cache.

sputhala-amd replied:
Inside `metadata_initialize_smi_tracks` we need to create separate tracks for each category.

If you meant about passing `gpu::is_vcn_activity_supported` as a serialized flag inside `serialize_gpu_metrics`, then I agree with that approach and that how I would like to do it too.
But I wanted the flag after the deserialization, when I inserted the data from `gpu_metrics_t` into DB. For serialization and deserialization this flag is not needed.

But if there are ideas about how this can be made better I am open to it.

mradosav-amd replied:
Right now, we are inserting tracks which are not used at all, I think tracks will be displayed only if some data is associated with it. 
Regarding second part and `gpu::is_vcn_activity_supported`, `rocpd_post_processing` will write data to database ether way, because those data carry 'same' information, VCN Activity. `rocpd_post_processing` should not care about from where data came, only should it write it to database or not. You can take a look at previous implementation and how `gpu::is_vcn_activity_supported` is embedded in serialized data. To be precise, in previous implementation you have:

``` 
serialize_xcp_metrics(gpu::is_vcn_activity_supported(m_dev_id), gpu::is_jpeg_activity_supported(m_dev_id), _gpu_metrics));
```

And depending on those arguments, we will change count of data which we are sending:
```
const bool   is_vcn_jpeg_supported = (use_vcn_activity || use_jpeg_activity);
const size_t chunk_count           = is_vcn_jpeg_supported ? 1 : xcp_count;
const size_t total_size = serialized_data_headers + (chunk_count * chunk_size);
```

`chunk_count` is set to `1`, if `gpu::is_vcn_activity_supported` or `gpu::is_jpeg_activity_supported` is supported, I guess you get the idea.

sputhala-amd replied:
Thanks for taking time and explaining this to me. I completely understand and agree to this.
I need to do some changes in order to accommodate what you are saying.

Currently I am passing data::gpu_metrics_t for serialization instead of the full structure amdsmi_gpu_metrics_t _gpu_metrics, since we have already processed this information inside the sample(). Hence I will need to modify the struct gpu_metrics_t to accommodate the design.

I will update the PR after testing the new fix. Thanks again !

mradosav-amd reviewed (changes requested):
Please address previous comments.

sputhala-amd committed:
Modify the gpu_metrics struct to accomodate XCP structure

sputhala-amd commented:
Hi @mradosav-amd You can review this PR again now. I made the following changes:

1. Refactor struct gpu_metrics_t (was xcp_metrics_t) to store VCN values separately for vcn_activity and xcp.vcn_busy, similarly for jpeg values. Before we were passing the whole amdsmi_gpu_metrics_t _gpu_metrics, so it was easy to serialize and deserialize the data and count sizes with AMDSMI_MAX_NUM_VCN, AMDSMI_MAX_NUM_JPEG, AMDSMI_MAX_NUM_XCP. However we have processed data from inside sample().

2. Serialization now take into consideration that xcp sizes can be different because of the UINT_MAX filter inside sample() during data storage.

Note for 1. & 2. I currently do this because we are doing the filtering of data for Perfetto and thought its better to use the same filtered data structure for serialization. If you think we are going to use trace_cache as the source for Perfetto in the future then I can keep the logic inside the serialize function to still use amdsmi_gpu_metrics_t _gpu_metrics.

3. Removed the setting values from being serialized again because the settings are already serialized and hence just accessing them during deserialization.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_post_processing.cpp:
```diff
@@ -405,71 +404,172 @@ rocpd_post_processing::get_pmc_event_with_sample_callback() const
 rocpd_post_processing::get_amd_smi_sample_callback() const
 {
-    struct xcp_metrics_t
+    struct gpu_metrics_t
```
This is better. Thanks a lot! Only thing I would suggest is to move `gpu_metrics_t` to some utility file without dependencies, and to reuse it inside `amd_smi` and `rocpd_post_processing`.  You can also move `serialize_gpu_metrics`, `deserialize_gpu_metrics` inside that file. In that way, we can keep things together and it would be easier to change it in the future. And we will benefit from separating it in sense of unit tests, because we can have a unit test which will serialize - deserialize struct and check if deserialize reconstructed struct correctly.

sputhala-amd replied:
I have addressed these comments and moved the struct and serialization logic to a new file.

dgaliffiAMD commented:
I rebased the branch to pick up the latest changes from `develop`.

sputhala-amd committed:
Add ctest automation for gpu interconnect metrics

sputhala-amd committed:
Refactor to move gpu_metrics struct and serialization to another file

dgaliffiAMD commented on projects/rocprofiler-systems/tests/rocprof-sys-gpu-connect-tests.cmake:
```diff
@@ -0,0 +1,93 @@
+        ARGS --counter-names "XGMI Read Data" "XGMI Write Data" -p
+    )
+
+    if(NOT skip_validation AND ${ENABLE_ROCPD_TEST} AND ${_VALID_GPU})
```
```suggestion
    if(${ENABLE_ROCPD_TEST} AND ${_VALID_GPU})
```
The "NOT skip_validation" check was redundant.

sputhala-amd replied:
Thanks David. There was also a CI failure on Ubuntu 24.04 • gfx950 due to a CI timeout:

2025-11-11T03:10:51.0065833 [rocprof-sys][145773][63][ensure_ci_timeout_backtrace] timeout after  **120.000 seconds**... Generating backtraces for 14 threads...
2025-11-11T03:10:51.0067158Z Backtrace:

I am turning off sampling by default to see if that can bypass the timeout error

dgaliffiAMD commented on projects/rocprofiler-systems/tests/rocprof-sys-gpu-connect-tests.cmake:
```diff
@@ -0,0 +1,93 @@
+    endif()
+endif()
+
+rocprofiler_systems_add_test(
```
Is it still worth running the test if "No valid transfers created"? Is it still collecting samples?

sputhala-amd replied:
Thats a good thought! 
Actually I am doing this just to add transferBench as a test in the ctest framework. Only with rocprofiler_systems_add_test() this gets added. 

In case someone wants to re-run with ctest -R transferBench, if the ctest are not added it says:
`No tests were found!!!`
Whereas in the current case it says:
```   
       Start 212: transferbench-sampling
1/1 Test #212: transferbench-sampling ...........***Skipped   0.85 sec
The following tests did not run:
        212 - transferbench-sampling (Skipped)
```

I think either option is fine. But the current option will skip the test.

dgaliffiAMD replied:
You could also use the "DISABLED" property here.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/rocprof-sys-gpu-connect-tests.cmake:
```diff
@@ -0,0 +1,93 @@
+    TARGET transferBench
+    GPU ON
+    ENVIRONMENT "${_base_environment};${_gpu_connect_environment}"
+    LABELS "transferbench"
```
```suggestion
    LABELS "transferbench;xgmi"
```

dgaliffiAMD commented on projects/rocprofiler-systems/tests/rocpd-validation-rules/gpu-connect/validation-rules.json:
```diff
@@ -0,0 +1,121 @@
+                    "comparison": "greater_than",
+                    "description": "Verify that 'rocm_hsa_api' appears in category at least 1500 times in table events_args",
+                    "error_message": "'rocm_hsa_api' category entries are fewer than expected in events_args",
+                    "expected_result": 100,
```
Description and expected do not match

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/gpu_metrics.cpp:
```diff
@@ -0,0 +1,330 @@
+    if(is_vcn_activity_supported) flags |= (1 << 0);  // Set bit for VCN activity support
+    if(is_jpeg_activity_supported)
+        flags |= (1 << 1);  // Set bit for JPEG activity support
+    serialize_uint8(result, flags);
```
```suggestion
    // Serialize capability flags
    struct {
        union {
            struct {
                uint8_t vcn_activity : 1;
                uint8_t jpeg_activity : 1;
            };
            uint8_t value;
        };
    } capabilities;

    capabilities.vcn_activity = is_vcn_activity_supported ? 1 : 0;
    capabilities.jpeg_activity = is_jpeg_activity_supported ? 1 : 0;
    serialize_uint8(result, capabilities.value);
```

If you want to make "flag setting" a little more readable, I suggest a structure like above.

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/gpu_metrics.hpp:
```diff
@@ -0,0 +1,126 @@
+ * Serialization format:
+ * 1. Support flags byte (1 byte):
+ *      - bit 0: is_vcn_activity_supported (device-level vs per-XCP)
+ *      - bit 1: is_jpeg_activity_supported (device-level vs per-XCP)
```
I have always thought that the "is_x_activity_supported" flags to be a little confusing. Is something like this easier to understand?
```suggestion
        // These flags determine how the activity information is provided in the data structures.
 *      - bit 0: vcn_is_device_level_only (device-level vs per-XCP) 
 *      - bit 1: jpeg_is_device_level_only (device-level vs per-XCP)
```

sputhala-amd committed:
Possible fix for timeout in CI

Fix redundant skip check in ctest
Add xgmi and pcie option in rocprof-sys-avail.

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/gpu_metrics.cpp:
```diff
@@ -0,0 +1,330 @@
+}
+
+void
+deserialize_gpu_metrics(const std::vector<uint8_t>& serialized_data,
```
I have a question. Do we need to handle the "is_vcn_supported" and "is_jpeg_supported" in the deserialization? Yes, it is required for serialization because the SMI is providing the data differently in for different GPUs. But, for deserialization, we are the consumers of the data. Can we make it easier on ourselves by populating the data the same regardless?  For example, use vcn_busy[0] and jpeg_busy[0].

sputhala-amd replied:
Not sure if I completely understand the question but the initial code was populating vcn_busy and jpeg_busy regardless of whether those values were xcp specific or just the activity values, however due to the processing of data that we do for UINT_MAX values for all the gpu_metrics it was impossible to know which values were returned as N/A and what was the size of each array for XCP. 
After deserialization these values needs to be populated on the right tracks and I could not find a good way to fix this case. For ex if the values after removing N/A values are:
XCP 0: [10, 20, 30]        (3 values)
XCP 1: [15, 25]            (2 values)
XCP 2: [5, 10, 15, 20]     (4 values)

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/amd_smi.cpp:
```diff
@@ -92,7 +92,7 @@ config_settings(const std::shared_ptr<settings>& _config)
         std::string, "ROCPROFSYS_AMD_SMI_METRICS",
         "amd-smi metrics to collect: " + default_metrics + jpeg_activity_support +
-            vcn_activity_support + ". " +
+            vcn_activity_support + " xgmi, pcie. " +
```
Do we need a similar "is supported" check for XGMI?

sputhala-amd replied:
We can do this "is_suuported" check. 

Do you think we need to do collectively for all the xgmi (and pcie) metrics like link_speed, link_width, read, write (so if at least one metric is available we say supported ) OR check for each xgmi metric individually.

dgaliffiAMD replied:
I was thinking a single `is_xgmi_supported` check. For example, on a typical Radeon desktop card `amd-smi xgmi` will return invalid data for all these metrics. In this case, we don't need to report it as a supported metric.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/rocpd-validation-rules/gpu-connect/validation-rules.json:
```diff
@@ -0,0 +1,121 @@
+{
+    "required_tables": [
```
Also, adding some validation of the hip_api category.

```suggestion
    "required_tables": [
        {
            "commit": "Validation rules for hip_api",
            "name": "events_args",
            "required_columns": [
                "event_id",
                "category",
                "stack_id",
                "parent_stack_id",
                "correlation_id"
            ],
            "validation_queries": [
                {
                    "comparison": "greater_than",
                    "description": "Verify that 'rocm_hip_api' appears in category at least 100 times in table events_args",
                    "error_message": "'rocm_hip_api' category entries are fewer than expected in events_args",
                    "expected_result": 100,
                    "query": "SELECT COUNT(*) FROM events_args WHERE category = 'rocm_hip_api';"
                },
                {
                    "comparison": "equals",
                    "description": "Check for missing category entries",
                    "error_message": "Empty or NULL category entries found in events_args",
                    "expected_result": 0,
                    "query": "SELECT COUNT(*) FROM events_args WHERE category IS NULL OR TRIM(category) = '';"
                }
            ]
        },
        {
            "commit": "Validation rules for hip_api",
            "name": "regions",
            "required_columns": [
                "id",
                "guid",
                "category",
                "name"
            ],
            "validation_queries": [
                {
                    "comparison": "greater_than",
                    "description": "Verify that 'rocm_hip_api' appears in category at least 50 times in table regions",
                    "error_message": "'rocm_hip_api' category entries are fewer than expected in regions",
                    "expected_result": 50,
                    "query": "SELECT COUNT(*) FROM regions WHERE category = 'rocm_hip_api';"
                },
                {
                    "comparison": "equals",
                    "description": "Ensure there are no HIP API calls that last 0 seconds",
                    "error_message": "Found HIP API captures where duration is 0",
                    "expected_result": 0,
                    "query": "SELECT COUNT(*) FROM regions WHERE category = 'rocm_hip_api' AND duration = 0;"
                },
                {
                    "comparison": "equals",
                    "description": "Check for any NULL values in the 'name' column of regions",
                    "error_message": "NULL entries found in the name column of regions",
                    "expected_result": 0,
                    "query": "SELECT COUNT(*) FROM regions WHERE name IS NULL;"
                }
            ]
        },
```

sputhala-amd committed:
Change2: Address review comments

Change ctest sampling to avoid timeout
Change variable name and code structuring

sputhala-amd committed:
Add option in ctest to run rocprof-sys-run without rewrite

Run transferbench with rocprof-sys-run without sampling

dgaliffiAMD commented:
Sorry, I pulled the latest from the branch, but now I'm seeing the following error in `validate-transpose-sampling-rocpd`.

```
Validating table: rocpd_pmc_event_*
✅ All required columns present in 'rocpd_pmc_event_781d461c8811500accbe357ea772a582': ['event_id', 'pmc_id', 'value']
✅ Row count check passed for 'rocpd_pmc_event_781d461c8811500accbe357ea772a582': 9080 rows (minimum: 2000)
❌ ERROR: Less than expected number of captured amd-smi-busy samples! (Table: 'rocpd_pmc_event_781d461c8811500accbe357ea772a582')
   Expected: greater_than 150, Got: 0
❌ ERROR: Less than expected number of captured amd-smi-temperature samples! (Table: 'rocpd_pmc_event_781d461c8811500accbe357ea772a582')
   Expected: greater_than 150, Got: 0
❌ ERROR: Less than expected number of captured amd-smi-power samples! (Table: 'rocpd_pmc_event_781d461c8811500accbe357ea772a582')
   Expected: greater_than 150, Got: 0
❌ ERROR: Less than expected number of captured amd-smi-memory-usage samples! (Table: 'rocpd_pmc_event_781d461c8811500accbe357ea772a582')
   Expected: greater_than 150, Got: 0
```

mradosav-amd reviewed (approved):
Thanks. Approving by my side.

sputhala-amd committed:
Change3: Fix sample insert bug and address review comments

xgmi and pci support check
renaming variables
additional hip_api validation in rocpd

sputhala-amd commented:
Debugging the hang seen on the ubuntu 24.04 - gfx950 system for CI tests.

dgaliffiAMD reviewed (changes requested):
Please add to the CHANGELOG.md too. Thank you.

dgaliffiAMD commented:
> Please add to the CHANGELOG.md too. Thank you.

I see it is part of https://github.com/ROCm/rocm-systems/pull/1798. Thank you.

sputhala-amd committed:
Reduce the load from the trnasferBench sample

The CI builds were timing out when flushing a big temporary file to the
DB: (2720824.23 KB / 2720.82 MB / 2.72 GB)...

dgaliffiAMD commented on projects/rocprofiler-systems/tests/rocprof-sys-gpu-connect-tests.cmake:
```diff
@@ -0,0 +1,96 @@
+# -------------------------------------------------------------------------------------- #
+
+set(_gpu_connect_environment
+    "ROCPROFSYS_ROCM_DOMAINS=hip_runtime_api,memory_copy,hsa_api"
```
```suggestion
    "ROCPROFSYS_ROCM_DOMAINS=hip_runtime_api,memory_copy"
```

dgaliffiAMD reviewed (commented):
You can try to remove the hsa_api samples to reduce the size of the output in your test.

sputhala-amd committed:
Merge branch 'develop' into xgmiPcieMetrics
